toolkit/tools/imagegen/configuration/users.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Parser for the image builder's Users configuration schema.
+
+package configuration
+
+import (
+	"fmt"
+	"strconv"
+)
+
+type User struct {
+	Name                string   `json:"Name"`
+	UID                 string   `json:"UID"`
+	PasswordHashed      bool     `json:"PasswordHashed"`
+	Password            string   `json:"Password"`
+	PasswordExpiresDays int64    `json:"PasswordExpiresDays"`
+	SSHPubKeyPaths      []string `json:"SSHPubKeyPaths"`
+	PrimaryGroup        string   `json:"PrimaryGroup"`
+	SecondaryGroups     []string `json:"SecondaryGroups"`
+	StartupCommand      string   `json:"StartupCommand"`
+}
+
+// IsValid returns an error if the User struct is not valid
+func (p *User) IsValid() (err error) {
+	err = p.NameIsValid()
+	if err != nil {
+		return
+	}
+	err = p.UIDIsValid()
+	if err != nil {
+		return
+	}
+	/* err = p.PasswordIsValid()
+	if err != nil {
+		return
+	} */
+	err = p.PasswordExpiresDaysIsValid()
+	if err != nil {
+		return
+	}
+	return
+}
+
+// NameIsValid returns an error if the User name is empty
+func (p *User) NameIsValid() (err error) {
+	if p.Name == "" {
+		return fmt.Errorf("invalid value for name (%s)", p.Name)
+	}
+	return
+}
+
+// UIDIsValid returns an error if the UID is outside bounds
+// Bounds can be checked using
+// $grep -E '^UID_MIN|^UID_MAX' /etc/login.defs
+func (p *User) UIDIsValid() (err error) {
+	const (
+		UID_lowerbound = 0 //root user
+		UID_upperbound = 60000
+	)
+	if p.UID != "" {
+		UID_i, _ := strconv.Atoi(p.UID)
+		if UID_i < UID_lowerbound || UID_i > UID_upperbound {
+			return fmt.Errorf("invalid value for UID (%s)", p.UID)
+		}
+	}
+	return
+}
+
+// This function does not account for the Password key being intentionally
+// unspecified versus set to blank
+// PasswordIsValid returns an error if the User password is empty
+/* func (p *User) PasswordIsValid() (err error) {
+	if p.Password == "" {
+		return fmt.Errorf("invalid value for Password (%s)", p.Password)
+	}
+	return
+} */
+
```

Comment:
@schmittjoseph and @PawelWMS, Any thoughts on how we should handle a user explicitly not setting the Password field versus intentionally setting it to blank?

Reply:
I no longer recall if we allowed users without passwords or not. If we didn't, then I'd treat both the same way - invalid. If we did, then there's no telling the difference without manually parsing the JSON. And we don't want that - it'll probably get messy. What we could do in such case, is add another `NoPassword` boolean field to the object. I do hope, however, that we simply don't allow blank passwords.

Reply:
Removing this function for now. Will resolve with a NoPassword field in a future PR (leading to a default behavior of not being able to log into that account).